Feature/v2 updates

fivetran-reneeli opened the pull request:
Pull Request
**Are you a current Fivetran customer?** 
<!--- Please tell us your name, title and company -->
Internal

**What change(s) does this PR introduce?** 
<!--- Describe what changes your PR introduces to the package and how to leverage this new feature. -->
We are updating the Salesforce package to improve the utility! This includes adding the new Contact Enhanced and Sales Velocity models, as well as updating the Opportunity Enhanced model. Finally we are adding integration with the Salesforce Formula package by embedding the macro outputs as part of our staging models so that your custom formula fields can be included.

**Did you update the CHANGELOG?** 
<!--- Please update the new package version’s CHANGELOG entry detailing the changes included in this PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes

**Does this PR introduce a breaking change?**
<!--- Does this PR introduce changes that will cause current package users' jobs to fail or require a `--full-refresh`? -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes (please provide breaking change details below.)
- [ ] No  (please provide an explanation as to how the change is non-breaking below.)


**Did you update the dbt_project.yml files with the version upgrade (please leverage standard semantic versioning)? (In both your main project and integration_tests)** 
<!--- The dbt_project.yml and the integration_tests/dbt_project.yml files contain the version number. Be sure to upgrade it accordingly -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes

**Is this PR in response to a previously created Bug or Feature Request**
<!--- If an Issue was created it is helpful to track the progress by linking it in the PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [] Yes, Issue/Feature [link bug/feature number here]
- [x] No 

**How did you test the PR changes?** 
<!--- Proof of testing is required in order for the PR to be approved. -->
<!--- To check a box, remove the space and insert an x in the box (eg. [x] CircleCi). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] CircleCi <!--- CircleCi testing is only applicable to Fivetran employees. --> 
- [ ] Local (please provide additional testing details below)

**Select which warehouse(s) were used to test the PR**
<!--- To check a warehouse remove the space and insert an x in the box (eg. [x] Bigquery). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] BigQuery
- [x] Redshift
- [x] Snowflake
- [x] Postgres
- [x] Databricks
- [] Other (provide details below)

**Provide an emoji that best describes your current mood**
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
:dancer:

**Feedback**

We are so excited you decided to contribute to the Fivetran community dbt package! We continue to work to improve the packages and would greatly appreciate your [feedback](https://www.surveymonkey.com/r/DQ7K7WW) on our existing dbt packages or what you'd like to see next.

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,35 +1,72 @@
+
+- This package enables users to:
+  - Understand the performance of your opportunities
+  - Understand what is going on in your sales funnel and 
```
Is this still true?

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,35 +1,72 @@
+  - Understand what is going on in your sales funnel and 
+  - Drill into how the members of your sales team are performing
+  - Have a daily view of sales activities 
+  - Access an enhanced contact list
```
"Access" sounds a bit strange in the context of the models we are providing. Can we change this to "leverage" instead?

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,35 +1,72 @@
-salesforce\_\_sales\_snapshot|A single row snapshot that provides various metrics about your sales funnel.
-salesforce\_\_opportunity\_enhanced|Each record represents an opportunity, enriched with related data about the account and opportunity owner.
-
+| [salesforce_manager_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__manager_performance)     |Each record represents a manager, enriched with data about their team's pipeline, bookings, losses, and win percentages.
```
```suggestion
| [salesforce__manager_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__manager_performance)     |Each record represents a manager, enriched with data about their team's pipeline, bookings, losses, and win percentages.
```

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,35 +1,72 @@
-salesforce\_\_opportunity\_enhanced|Each record represents an opportunity, enriched with related data about the account and opportunity owner.
-
+| [salesforce_manager_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__manager_performance)     |Each record represents a manager, enriched with data about their team's pipeline, bookings, losses, and win percentages.
+| [salesforce_owner_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__owner_performance)         |Each record represents an individual member of the sales team, enriched with data about their pipeline, bookings, losses, and win percentages.
```
```suggestion
| [salesforce__owner_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__owner_performance)         |Each record represents an individual member of the sales team, enriched with data about their pipeline, bookings, losses, and win percentages.
```

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,35 +1,72 @@
-
+| [salesforce_manager_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__manager_performance)     |Each record represents a manager, enriched with data about their team's pipeline, bookings, losses, and win percentages.
+| [salesforce_owner_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__owner_performance)         |Each record represents an individual member of the sales team, enriched with data about their pipeline, bookings, losses, and win percentages.
+| [salesforce_sales_snapshot](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__sales_snapshot)               |A single row snapshot that provides various metrics about your sales funnel.
```
```suggestion
| [salesforce__sales_snapshot](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__sales_snapshot)               |A single row snapshot that provides various metrics about your sales funnel.
```

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,35 +1,72 @@
+| [salesforce__contact_enhanced](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__contact_enhanced)  |Each record represents a contact with additional account and owner information.
+| [salesforce__daily_activity](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__daily_activity)  |Each record represents a daily summary of the number of sales activities, for example tasks and opportunities closed.
+
+# 🤔 Who is the target user of this dbt package?
+- You use Fivetran's [Salesforce connector](https://fivetran.com/docs/applications/salesforce)
+- You use dbt
+- You want a staging layer that cleans, tests, and prepares your Salesforce data for analysis as well as leverage the analysis ready models outlined above.
```
Same comment from the source package, this is already covered in the above section. We can remove it entirely.

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -38,11 +75,12 @@ By default, this package looks for your Salesforce data in the `salesforce` sche
+    salesforce_database: your_database_name    
     salesforce_schema: your_schema_name
-    salesforce_database: your_database_name
+    salesforce__<default_source_table_name>_identifier: your_table_name
```
Since you already included this in the optional section below, we should remove it here.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks for opening this PR and applying these changes! I do have a few inline comments you can see below that we will want to address, and I also have a few other notes:
- [ ] Adjust the release version to no longer by v1.0.0. I am not getting the feeling that this is a v1 release just yet and I would like to evaluate with the team what our standards for this should be going forward.
- [ ] The README does not follow the same standards you applied to the dbt_salesforce_source updates you made. Please refer to my notes and the structure of the source README and make the appropriate changes. A good rule of thumb is that we want the same content in the source README to also be in the transform README. This way customers don't have to flip between repos. Mainly, I noticed you didn't include the Salesforce Formula Utils integration notes.
- [ ] Can we change the naming convention of `salesforce__date_spine` and `salesforce__opportunity_aggregates` to include `int_` at the start. Once materialized, these models look like final output models. We will want to adjust that accordingly.
- [ ] It looks like you didn't update the CHANGELOG with the updates from this PR. There are a lot. So we will want to document all breaking and notable changes.
- [ ] Please take a look at the issue with the user_role join to the opportunity enhanced model. This join doesn't work and I think it will be beneficial to create an intermediate model to join user and user_role before bringing it into this model.
- [ ] With us not being able to generate the `sales_velocity` model, I think it may be beneficial to at least leverage the opportunity_line and product_2 staging models and generate a `salesforce__opportunity_line_enhanced` model. This would be a fairly easy join and provide a lot of rich information for users to get a jump start on some deeper opportunity analysis. Let's add this new model.

fivetran-joemarkiewicz commented on models/salesforce__opportunity_enhanced.sql:
```diff
@@ -2,66 +2,93 @@ with opportunity as (
+
+    -- If using user_role table, the following will be included, otherwise it will not.
+    {% if var('salesforce__user_role_enabled', True) %}
+    left join user_role 
+        on opportunity_owner.user_id = user_role.user_role_id
```
Ignore the intermediate logic above, we just need to. change this to be user_role_id to user_role_id

fivetran-reneeli commented on README.md:
```diff
@@ -38,11 +75,12 @@ By default, this package looks for your Salesforce data in the `salesforce` sche
-
-This package allows users to add additional columns to the opportunity enhanced table. Columns passed through must be present in the downstream source account table or user table. If you want to include a column from the user table, you must specify if you want it to be a field relate to the opportunity_manager or opportunity_owner.
+### Adding Passthrough Columns
+This package allows users to add additional columns to the `opportunity enhanced` model and `contact enhanced` model. Columns passed through must be present in the downstream source account, contact, or user table. If you want to include a column from the user table, you must specify if you want it to be a field relate to the opportunity_manager or opportunity_owner.
```
In the `opportunity_enhanced` model, the user table gets passed in twice, once as opportunity_manger and once as opportunity_owner. So a customer passing in a custom field from any of tables that `opportunity_enhanced` depends on (account, user, or user_role) must add it to the `opportunity_enhanced_pass_through_columns` as well, but for fields specifically coming from user, it needs to be specified either as **opportunity_manager**.user_custom_column_1 or **opportunity_owner**.user_custom_column_1

fivetran-reneeli commented:
Ready for review. In addition I added enable/disable configs for tables that may not be used by all customers (used a cutoff of 85% usage among our customers) in the source package and in the downstream models. Followingly, I added an extra dbt run to test out those configs in integration testing.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+# dbt_salesforce 0.7.0
+🎉 Salesforce Package Updates 🎉
+
+We are updating the Salesforce package! To improve its utility, the changes include the following:
```
Can you reference the PR here and make a blanket statement that this PR involves the following changes.
```suggestion
PR [#30](https://github.com/fivetran/dbt_salesforce/pull/30) includes various updates to the Salesforce package! To improve its utility, the changes include the following:
```

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+
+We are updating the Salesforce package! To improve its utility, the changes include the following:
+## Features
+  - Creating a new Contact Enhanced, Daily Activity, and Opportunity Line Item Enhanced model as well as updating the Opportunity Enhanced model.
```
Minor request, can you add the model names to match what the user will see in their warehouse. So "Contact Enhanced" should be `salesforce__contact_enhanced`.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+🎉 Salesforce Package Updates 🎉
+
+We are updating the Salesforce package! To improve its utility, the changes include the following:
+## Features
```
Can you also add any new variables that you updated the README.

fivetran-joemarkiewicz replied:
I would also include any new variables that were added. If you want to just link to the section of the README that details these variables then I think that would be good as well.

fivetran-joemarkiewicz commented on models/salesforce__daily_activity.sql:
```diff
@@ -0,0 +1,105 @@
+    
+    select 
+        {{ dbt_utils.date_trunc('day', 'activity_date') }} as activity_date,
+        count(task_id) as tasks
```
Bump on this comment that was missed.

fivetran-joemarkiewicz commented on models/salesforce__daily_activity.sql:
```diff
@@ -0,0 +1,105 @@
+
+    select 
+        {{ dbt_utils.date_trunc('day', 'activity_date') }} as activity_date,
+        count(event_id) as events
```
Bump on this comment that was missed.

fivetran-joemarkiewicz commented on models/salesforce__daily_activity.sql:
```diff
@@ -0,0 +1,105 @@
+opportunities_closed as (
+    select
+        close_date,
+        count(case when status = 'Won' then opportunity_id else null end) as won_opportunities,
+        count(case when status = 'Lost' then opportunity_id else null end) as lost_opportunities
```
Bump on this comment that was missed.

fivetran-joemarkiewicz commented on models/salesforce__opportunity_enhanced.sql:
```diff
@@ -2,66 +2,93 @@ with opportunity as (
+
+        --The below script allows for pass through columns.
+
+        {% if var('opportunity_enhanced_pass_through_columns',[]) != [] %}
```
Bump on this comment that was missed.

fivetran-reneeli replied:
The `opportunity_enhanced_pass_through_columns` includes all pass through columns coming from the opportunity staging model, in addition to account, user, and user_role. I believe it's done this way because `user` gets joined twice in this model, so doing this 'mega' pass through column variable allows the user to specify whether field is coming from user1 or user2. 

```
 If you want to include a column from the `user` table, you must specify if you want it to be a field related to the opportunity_manager or opportunity_owner.
 
   opportunity_enhanced_pass_through_columns: [opportunity_manager.user_custom_column_1]
```

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli this is looking good and so close to being ready for release! A few things I would like to be addressed before approval:
- [ ] A few minor requested changes within the inline comments of the CHANGELOG
- [x] It looks like a few comments from the last review were missed. Especially within the `salesforce__daily_activity` model. Would you be able to address those comments.
- [x] Similarly it seems the review notes from the previous review within the README still need to be addressed. Can you please address those and make any necessary changes.
- [ ] I noticed there are a few fields within the dbt docs that are not documented. Would you be able to go through and make sure all the base fields have a corresponding definition.

fivetran-joemarkiewicz commented on README.md:
````diff
@@ -65,55 +112,93 @@ vars:
+```
 
+### Adding Passthrough Columns
+This package allows users to add additional columns to the `opportunity enhanced` model and `contact enhanced` model. - For the `opportunity enhanced` model, columns passed through in `opportunity_enhanced_pass_through_columns` must also be present in the upstream source `opportunity`, `account`, `user`, or `user_role` table. If you want to include a column from the `user` table, you must specify ...
````
```suggestion
This package allows users to add additional columns to the `opportunity enhanced` model and `contact enhanced` model. For the `opportunity enhanced` model, columns passed through in `opportunity_enhanced_pass_through_columns` must also be present in the upstream source `opportunity`, `account`, `user`, or `user_role` table. If you want to include a column from the `user` table, you must specify if you want it to be a field related to the opportunity_manager or opportunity_owner.
```

fivetran-joemarkiewicz commented on README.md:
````diff
@@ -65,55 +112,93 @@ vars:
 
-## Contributions
+Include the following within your `dbt_project.yml` file:
+```yml
+# Using the opportunity source table as example, update the opportunity variable to reference your newly created model that contains the formula fields:
+  opportunity: "{{ ref('my_opportunity_formula_table') }}"
````
```suggestion
```yml
# Using the opportunity source table as example, update the opportunity variable to reference your newly created model that contains the formula fields:
vars:
  opportunity: "{{ ref('my_opportunity_formula_table') }}"
```

fivetran-joemarkiewicz commented on packages.yml:
```diff
@@ -1,3 +1,7 @@
 packages:
- - package: fivetran/salesforce_source
-   version: [">=0.4.0", "<0.5.0"]
+#  - package: fivetran/salesforce_source
+#    version: [">=0.5.0", "<0.6.0"]
+
+    - git: https://github.com/fivetran/dbt_salesforce_source.git
+      revision: feature/package_updates_v2
+      warn-unpinned: false
```
Bump to make sure we switch before merge.

fivetran-joemarkiewicz commented on dbt_project.yml:
```diff
@@ -13,11 +13,35 @@ vars:
+    order: "{{ ref('stg_salesforce__order') }}"
+    opportunity_line_item: "{{ ref('stg_salesforce__opportunity_line_item') }}"
 
+    # Salesforce pass through column variables
+    account_pass_through_columns: []
```
```suggestion
    # Salesforce pass through column variables
    opportunity_enhanced_pass_through_columns: []
    account_pass_through_columns: []
```

fivetran-joemarkiewicz reviewed (approved):
@fivetran-reneeli thanks for addressing all my review notes! I have a few minor notes in this review that I would like for you to address (most of them README related) before you merge this. Once you make those changes this will be good to go!

Thanks and let's coordinate release once you have the source PR merged 😄
